include/regs.h

```diff
@@ -0,0 +1,273 @@
+        : [v] "+r" (value), [m] "+r" (mask)                                    \
+    )
+
+#define CLEAR_BITS_BY_MASK_FUNC(value, bitmask)                                \
+    uint64_t val = value;                                                      \
+    uint64_t mask = bitmask;                                                   \
+    CLEAR_BITS_BY_MASK(val, mask);                                             \
+    return val;
+
+#define IS_BIT_ENABLED(var, result, bit_position)                              \
+    __asm__ __volatile__(                                                      \
+        "mov %[res], %["#var"]\n"                                              \
+        "lsr %[res], %[res], #"#bit_position"\n"                               \
+        "and %[res], %[res], #1\n"                                             \
+        : [res] "=r" (result) : [var] "r" (var)                                \
+    )
+
+#define IS_BIT_ENABLED_FUNC(var, bit_position)                                 \
+    uint64_t result;                                                           \
+    IS_BIT_ENABLED(var, result, bit_position);                                 \
+    return result;
+
+#define IS_BIT_DISABLED(var, result, bit_position)                             \
+    __asm__ __volatile__(                                                      \
+        "mov %[res], %["#var"]\n"                                              \
+        "lsr %[res], %[res], #"#bit_position"\n"                               \
+        "mvn %[res], %[res]\n"                                                 \
+        "and %[res], %[res], #1\n"                                             \
+        : [res] "=r" (result) : [var] "r" (var)                                \
+    )
+
+#define IS_BIT_DISABLED_FUNC(var, bit_position)                                \
+    uint64_t result;                                                           \
+    IS_BIT_DISABLED(var, result, bit_position);                                \
+    return result;
+
+// -----------------------------------------------------------------------------
+// System Registers
+// -----------------------------------------------------------------------------
+
+#define GET_SYSREG(sysreg, dest)                                               \
+    __asm__ __volatile__(                                                      \
+        "mrs %0, "#sysreg"\n"                                                  \
+        : "=r"(dest)                                                           \
+    )
+
+#define GET_SYSREG_FUNC(reg)                                                   \
+    uint64_t val;                                                              \
+    GET_SYSREG(reg, val);                                                      \
+    return val;
+
+#define GET_SYSREG_FIELD(sysreg, dest, mask, lsb)                              \
```

Comment:
I still think the bit manipulation should be done in C/++ and assembly only used for cases where it's actually necessary (e.g. `mrs`), but esp. given I said I'd contribute a fix for that and still haven't, I've got no problem approving the PR with this and can always submit an improvement later.

Reply:
I agree, there's probably many efficiencies to be gained in this file. Lets keep this implementation as the starting point (cause it works and has test coverage), and use some RFCs in the future to discuss and propose improvements! I would definitely be excited about getting rid of as much assembly that we have to maintain as possible.